yarn-project/simulator/src/avm/opcodes/addressing_mode.ts

```diff
@@ -3,26 +3,28 @@ import { strict as assert } from 'assert';
 import { type TaggedMemoryInterface } from '../avm_memory_types.js';
 
 export enum AddressingMode {
-  DIRECT,
-  INDIRECT,
-  INDIRECT_PLUS_CONSTANT, // Not implemented yet.
+  DIRECT = 0,
+  INDIRECT = 1,
+  RELATIVE = 2,
+  INDIRECT_RELATIVE = 3,
 }
 
 /** A class to represent the addressing mode of an instruction. */
 export class Addressing {
   public constructor(
     /** The addressing mode for each operand. The length of this array is the number of operands of the instruction. */
     private readonly modePerOperand: AddressingMode[],
-  ) {
-    assert(modePerOperand.length <= 8, 'At most 8 operands are supported');
-  }
+  ) {}
 
-  public static fromWire(wireModes: number): Addressing {
+  // TODO(facundo): 8 for backwards compatibility.
+  public static fromWire(wireModes: number, numOperands: number = 8): Addressing {
     // The modes are stored in the wire format as a byte, with each bit representing the mode for an operand.
     // The least significant bit represents the zeroth operand, and the most significant bit represents the last operand.
-    const modes = new Array<AddressingMode>(8);
-    for (let i = 0; i < 8; i++) {
-      modes[i] = (wireModes & (1 << i)) === 0 ? AddressingMode.DIRECT : AddressingMode.INDIRECT;
+    const modes = new Array<AddressingMode>(numOperands);
+    for (let i = 0; i < numOperands; i++) {
+      modes[i] =
+        (((wireModes >> i) & 1) * AddressingMode.INDIRECT) |
+        (((wireModes >> (i + numOperands)) & 1) * AddressingMode.RELATIVE);
     }
```

Comment:
What is the `+ numOperands` doing here? Is that basically saying "'relative' bits start at bit index `numOperands`"? Does this need a special case for instructions like `*CALL` that have >4 operands and won't be able to enable relative addressing (at least not for all operands)?

Reply:
Yes that what it means!
> Does this need a special case for instructions like *CALL that have >4

It does not need special casing here, we just need to make the wire format indirect be a u16 for those operations to be usable in relative mode. Also mentioned here: https://docs.google.com/document/d/1uiSp8Q2Pu8L2UCba1PPPNQfxxK3d78_-vrzud7PECnw/edit#bookmark=id.wmo1rdovzz9f